### PR TITLE
HFModelPusher component proposal

### PR DESCRIPTION
**Pull request by deep-diver**
Wrote a HFModelPusher component proposal (cc: @sayakpaul as the co-developer)

### CODE REVIEW

File: proposals/20220823-huggingface_model_pusher.md
Comment (sayakpaul) [reply proposals/20220823-huggingface_model_pusher.md:54]: If it would check out a new branch, how is the end user supposed to use the model using `from_pretrained()`? 

I know HF Hub can be used as a generic tool for hosting models of any type. But for an HF user, it would be a natural expectation to use the model checkpoints pushed to the Hub using APIs like `from_pretrained()`. We need to clarify this. 

How will we handle the README.md of the underlying Hub repository? Here's how HF does it: 

https://github.com/huggingface/transformers/blob/f0d496828d3da3bf1e3c8fbed394d7847e839fa6/src/transformers/trainer.py#L3274

**Review by sayakpaul — CHANGES_REQUESTED** [review]

File: proposals/20220823-huggingface_model_pusher.md
Comment (deep-diver) [reply proposals/20220823-huggingface_model_pusher.md:54]: > If it would check out a new branch, how is the end user supposed to use the model using from_pretrained()?

Since the model is pushed for experimental purpose, I think it would be good to track the versions of the model within separate branches. When the model is ready to be open to public, one can manually merge the right version(branch) into the `main` branch. 

> How will we handle the README.md of the underlying Hub repository? Here's how HF does it:

I have not thought about the model card. Should we need it? It looks like `create_model_card()` is useful if the model has some dependencies to the other HuggingFace features such as `Datasets`

Maybe we can prepare a template `README.md`, then inject some values into it, then add it to the commit. Or we can add this feature after the #82 (Model Card Generator) is fully developed. WDYT?

Comment (sayakpaul) [reply proposals/20220823-huggingface_model_pusher.md:54]: > Since the model is pushed for experimental purpose, I think it would be good to track the versions of the model within separate branches. When the model is ready to be open to public, one can manually merge the right version(branch) into the main branch.

Let's add this info to the proposal. 

> I have not thought about the model card. Should we need it? It looks like create_model_card() is useful if the model has some dependencies to the other HuggingFace features such as Datasets

Maybe we can prepare a template README.md, then inject some values into it, then add it to the commit. Or we can add this feature after the https://github.com/tensorflow/tfx-addons/issues/82 (Model Card Generator) is fully developed. WDYT?

All good.

File: proposals/20220823-huggingface_model_pusher.md
Comment (deep-diver) [reply proposals/20220823-huggingface_model_pusher.md:49]: Greate idea!

File: proposals/20220823-huggingface_model_pusher.md
Comment (deep-diver) [reply proposals/20220823-huggingface_model_pusher.md:52]: oops! my bad, will fix it soon

File: proposals/20220823-huggingface_model_pusher.md
Comment (deep-diver) [reply proposals/20220823-huggingface_model_pusher.md]: > Let's add this info to the proposal.

Absolutely! Thanks for the good one

**Comment by deep-diver** [comment]
addressed the comments. 

need reviews from @casassg, @rcrowe-google

**Review by sayakpaul — APPROVED** [review]
LGTM.

File: proposals/20220823-huggingface_model_pusher.md
Comment (codesue) [inline proposals/20220823-huggingface_model_pusher.md:28]: It might be nice to include a README and/or a model_card_metadata [config](https://github.com/huggingface/hub-docs/blob/main/modelcard.md?plain=1) as inputs for additional documentation and discoverability.

Comment (deep-diver) [reply proposals/20220823-huggingface_model_pusher.md:28]: @codesue 

Thank you for the suggestion! me and @sayakpaul had the same thought. 

Specifically, it would be great to upload a model card generated by `Evaluator` and [`model-card-toolkit`](https://github.com/tensorflow/model-card-toolkit). However the `model-card-toolkit` is a on-going project to be ported into TFX Add-on, so I thought maybe upgrade `HFModelPusher` when `model-card-toolkit` is completed.

By the way, your suggestion on the [model_card_metadata config](https://github.com/huggingface/hub-docs/blob/main/modelcard.md?plain=1) sounds good too!  But, it has many many information to fill in, so it would be inappropriate for a TFX component to fill in automatically. Do you have any idea how to make things easier for users to use this?

Comment (casassg) [reply proposals/20220823-huggingface_model_pusher.md:28]: Similar to how mct fills information, it may be good to automatically fill some of this by adding Statistics and such as inputs to the component. See the existing ModeCardGenerator component for reference -> https://github.com/tensorflow/model-card-toolkit/blob/master/model_card_toolkit/tfx/executor.py

Comment (deep-diver) [reply proposals/20220823-huggingface_model_pusher.md:28]: so, taking outputs from `Evaluator` and `StatisticsGen` optionally.
- if both Artifacts exist, create a `ModelCardToolkit`
- then fill some information from it

File: proposals/20220823-huggingface_model_pusher.md
Comment (deep-diver) [reply proposals/20220823-huggingface_model_pusher.md:26]: Right, i agree

Comment (sayakpaul) [reply proposals/20220823-huggingface_model_pusher.md:26]: It does not _need_ to. But having it inherited from `Pusher` will be beneficial. I guess that's what @deep-diver meant.

File: proposals/20220823-huggingface_model_pusher.md
Comment (codesue) [reply proposals/20220823-huggingface_model_pusher.md:28]: Yeah, that sounds good! Not all the HF model card fields are available in MCT, and MCT doesn't support custom fields yet, so we won't be able to completely fill in the HF config section automatically, but there's still a lot that can be automatically annotated if we use a custom template.

Comment (deep-diver) [reply proposals/20220823-huggingface_model_pusher.md:28]: The outputs of `Trainer` doesn’t have metrics information, and evaluation is done with `Evaluator`. 

So, we can include very general information such as display name, tags, which dataset was used, or something like it. We can prepare a template file filled up with default values, then let them overwrite if any inputs are given.

Comment (deep-diver) [reply proposals/20220823-huggingface_model_pusher.md:28]: I think [`modelcards`](https://github.com/nateraw/modelcards) package is useful for this purpose.

Comment (deep-diver) [reply proposals/20220823-huggingface_model_pusher.md:28]: @sayakpaul @codesue @casassg 

Or it would be easier to just put HTML contents generated by MCT into the markdown model card in HuggingFace Mode Repo. WDYT?

Comment (sayakpaul) [reply proposals/20220823-huggingface_model_pusher.md:28]: I think we would want to avoid the code for turning an HTML page into a separate markdown file for the Hugging Face Hub README. IMO, we could develop separate utilities for this purpose:

* Have a template for the model card (TFX and HF have their own formats, it seems, but there's overlap in information).
* Have a utility that generates HTML page from the populated template text 
* Have utility that generates a markdown file from the populated template text

Comment (deep-diver) [reply proposals/20220823-huggingface_model_pusher.md:28]: Right. 

Since this component could be over complicated if we include Model Card generation feature, I thought two possible solutions:
1. Create another custom TFX evaluation component for HuggingFace Model 
2. Or just run `evaluate` inside this component(`HFModelPusher`) over the test dataset and fill evaluation results into the Model Card
3. Or leverage the existing `Evaluator` standard TFX component

**Review by rcrowe-google — APPROVED** [review]
/lgtm

**Comment by casassg** [comment]
it will need a /merge from Robert as he is the sole owner XD

File: proposals/20220823-huggingface_model_pusher.md
Comment (codesue) [reply proposals/20220823-huggingface_model_pusher.md:28]: Model Card Toolkit can also create markdown or any arbitrary file types since you can have custom templates and filenames, so it wouldn't require converting HTML to markdown. Anyway, I'm in favor of your and @sayakpaul's original idea to upgrade `HFModelPusher` to include a model card later in order to prevent increasing scope and to avoid adding dependencies and unknowns.

**Comment by deep-diver** [comment]
@casassg 

Once this PR is merged, I will make another PR to update this proposal to include HuggingFace Space Deployment feature. Or should I do it now? I want to do it on a separate PR to clearly discuss different concerns.

**Comment by casassg** [comment]
I think that's fine. I can't merge this due to Robert being owner of the proposals himself (I could merge it but prefer to let him do this as I dont want to force merge either) You may want to tag him on slack next week to double check

**Comment by deep-diver** [comment]
then I will wait until @rcrowe-google merges this PR :)

**Comment by casassg** [comment]
Given Robert is out for a couple days, and given he approved this. I will manually merge, to help work continue on this

**Comment by rcrowe-google** [comment]
Sorry I missed this, and thanks for merging!

**Pull request merged**
